### PR TITLE
Redesign around pipelines

### PR DESCRIPTION
**Pull request by gsganden**
Redesigns Creevey around a `Pipeline` abstraction. Clarifies Creevey's purpose: it is a library for easily processing data in files, including but not limited to image files. Greatly simplifies code for downloading and resizing and makes it much easier to add new operations by handling currency in `Pipelines` rather than having to handle it separately for each function.

Functionality to do grouped train-test splits and create imagenet-style symlinks is being removed from Creevey because it no longer fits. I would like to get away from using directory structure to encode labels anyway. `group_train_test_split` is useful but not difficult to recreate elsewhere.

`Dataset` classes are also being removed, because `Pipelines` are better for handling the hard parts of the problems that `Dataset` classes are meant to address.

I merged into WIP without a code review by mistake, but we can still do a review before merging into master.

I'd suggest reading the README and running the code in it, and then reviewing `creevey/pipelines/core.py` before moving on to the other files.

Pull Request Checklist
 - [X] Pull request is either from WIP to master or from some other branch to WIP.
 - [X] Pull request includes a description of the change and the reason behind it.
 - [X] Pull request includes unit tests for any new functionality. 
 - [X] `pytest` passes locally.
 - [X] `black creevey tests --skip-string-normalization` has been used to format the code.
 - [X] `flake8 creevey tests` passes locally.
 - [X] CHANGELOG has been updated.
 - [X] Version in `_version.py` has been updated.
 - [X] README and docs have been updated (if applicable).

### CODE REVIEW

File: creevey/pipelines/core.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+            else:
+                try:
+                    stage = self.load_func(inpath)
+                    for op in self.ops:
+                        stage = op(stage)
+                    self.write_func(stage, outpath)
+                except exceptions_to_catch as e:
+                    logging.error(e, inpath)
+
+        return pipeline_func
```
Comment (jscottcronin) [inline creevey/pipelines/core.py]: I think you can remove the `_build_pipeline_func` abstraction and instead just have a method on the class called `pipeline_func`. This will automatically add the `self.pipeline_func` you have in the attributes. Here is what I mean:
```python
def __init__(
    self,
    load_func: Callable[[PathOrStr], Any],
    ops: Optional[Iterable[Callable[[Any], Any]]],
    write_func: Callable[[Any, PathOrStr], None],
) -> None:
    """
    Compose the provided functions, and store them as attributes.
    Store `load_func`, `ops`, and `write_func` as attributes with
    the corresponding names. Create an additional attribute
    `pipeline_func` that composes those functions for when `run`
    is called.
    See the `Pipeline` docstring for information about the form that
    `load_func`, `ops`, and `write_func` are expected to take.
    """
    self.load_func = load_func
    self.ops = ops if ops is not None else []
    self.write_func = write_func


def pipeline_func(
    inpath: PathOrStr,
    outpath_func: PathOrStr,
    skip_existing: bool,
    exceptions_to_catch: Optional[Tuple[Exception]] = None,
):
    outpath = outpath_func(inpath)
    if skip_existing and Path(outpath).is_file():
        logging.warning(
            f'Skipping {inpath} because there is already a file at corresponding '
            f'output path {outpath}'
        )
    else:
        try:
            stage = self.load_func(inpath)
            for op in self.ops:
                stage = op(stage)
            self.write_func(stage, outpath)
        except exceptions_to_catch as e:
            logging.error(e, inpath)
```

File: creevey/pipelines/core.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+        exceptions_to_catch
+            Tuple of exception types to catch. An exception of one of
+            these types will be logged with logging level ERROR and the
+            relevant file will be skipped.
+        """
+        Parallel(n_jobs=n_jobs, prefer='threads')(
+            delayed(self.pipeline_func)(
+                path, path_func, skip_existing, exceptions_to_catch
+            )
+            for path in tqdm(inpaths)
```
Comment (jscottcronin) [inline creevey/pipelines/core.py]: Does tqdm work magically in parallelized functions? I recall this did work well in the prior version of creevey, but I'm confused how it is able to keep track. Do you know?

File: creevey/load_funcs/load_funcs.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        URL of file to download
+    """
+    session = _get_session()
+    response = session.get(url)
+    _check_status(url, response)
+    return response
+
+
+def _get_session():
+    if getattr(threadLocal, 'session', None) is None:
```
Comment (jscottcronin) [inline creevey/load_funcs/load_funcs.py]: I didn't know you could do this. Really nice!

File: creevey/load_funcs/image.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    -------
+    Image as NumPy array
+    """
+    response = get_response(url)
+    image = _load_image_from_response(response)
+    return image
+
+
+def _load_image_from_response(response):
+    image = np.asarray(bytearray(response.content), dtype="uint8")
```
Comment (jscottcronin) [inline creevey/load_funcs/image.py]: OooOoO, I've always thought we had to create the bytesIO object. This seems so much cleaner! You have found some really nice code snippets to build into this library!

File: creevey/load_funcs/image.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    image = cv.imdecode(image, -1)  # load as-is, e.g. including alpha channel
```
Comment (jscottcronin) [inline creevey/load_funcs/image.py]: the explanation here did not clarify to me why this is needed. I looked up this function, and still do not get the purpose. Can you clarify for me?

File: creevey/load_funcs/image.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    return image
+
+
+def load_image(path: PathOrStr) -> np.array:
```
Comment (jscottcronin) [inline creevey/load_funcs/image.py]: Naming convention wise, I would have expected load_image to be `download_image` followed by `_load_image_from _response`.  Only reading docstring did I realize we load from disk. Is there a reason we do not want a function here that combines the two above. Maybe `download_image_into_memory` or something along that line?

File: creevey/ops/image.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    _validate_resize_inputs(shape, min_dim)
+    if min_dim is not None:
+        shape = _find_min_dim_shape(image, min_dim)
+    resized = cv.resize(image, dsize=shape[::-1])
+    return resized
+
+
+def _validate_resize_inputs(shape, min_dim) -> None:
+    if (shape is None) + (min_dim is None) == 1:
+        pass
```
Comment (jscottcronin) [inline creevey/ops/image.py]: very clever!

File: creevey/__init__.py [hunk 8]
```diff
@@ -1,10 +1,2 @@
...
-    download_files,
-    download_images_as_png,
-    download_single_file,
-)
-from creevey.model_selection import group_train_test_split
-from creevey.resize import resize_image, resize_file, resize_multiple_files
-from creevey.symlink import create_imagenet_style_symlinks
-from creevey.dataset import S3Downloader, S3TarfileDataset, TarfileExtractor
+from ._version import __version__
+from .pipelines import *  # noqa: F401, F403
```
[the remark this answers is not in the document]
Comment (gsganden) [reply creevey/__init__.py]: My understanding is that `from x import *` does **not** import import items that have been marked as private, and I have confirmed that e.g. `from creevey.load_funcs import get_response` works while `from creevey.load_funcs import _get_session` does not.

I know that `import *` is generally frowned upon, but I see only upsides in using it in an `__init__.py` file where you really do just want to bring all of the public stuff from a submodule into the module's namespace.

File: creevey/constants.py [hunk 9]
```diff
@@ -0,0 +1,4 @@
+from pathlib import Path
+from typing import Union
+
+PathOrStr = Union[Path, str]
```
[the remark this answers is not in the document]
Comment (gsganden) [reply creevey/constants.py]: I think this approach not that unusual for people who use type hints -- see e.g. https://realpython.com/python-type-checking/#type-aliases -- and I think it aids readability once you get over the novelty.

File: creevey/pipelines/core.py (hunk 1 above)
Comment (gsganden) [reply creevey/pipelines/core.py]: Yep, good catch. I make this change in the PR that Michael reviewed. I'd like to leave it as-is for this PR and just fix it in the next one to save myself a few steps. (That PR has been approved, so it will not be far behind this one.)

File: creevey/load_funcs/image.py (hunk 5 above)
Comment (gsganden) [reply creevey/load_funcs/image.py]: Ah, the joys of OpenCV. `cv.imdecode` takes a numeric code that tells it whether to read the image as grayscale (0), 3-channel (>0), or as-is including alpha channel (-1). Here I am using -1 to read the image as-is.

File: creevey/load_funcs/image.py (hunk 6 above)
Comment (gsganden) [reply creevey/load_funcs/image.py]: The idea is that a Creevey pipeline starts with a `load_func` that loads an item into memory from a specified location (which can be e.g. a URL or a location on disk) and ends with a `write_func` that writes out an item to a specified location (which can be e.g. an S3 URI or a location on disk). `download_image` and `load_image` are both `load_func`s, so they load an image into memory from a URL and from a location on disk, respectively. I will revise the names for clarity.

If you want to download an image and write to disk and then do further processing and presumably write to disk again, you have two options:

1. Run two pipelines in succession, where the first downloads and writes to disk, and the second loads from disk and goes from there.
2. Create a `load_func` that writes to disk as a side effect.

File: creevey/load_funcs/image.py (hunk 5 above)
Comment (gsganden) [reply creevey/load_funcs/image.py]: Maybe I'll add a line that gives a descriptive name to the magic number `-1`, like so:

```
    load_all_channels_code = -1
    image = cv.imdecode(image, flags=load_all_channels_code)  # load all channels, including an alpha channel if present
```

File: creevey/load_funcs/image.py (hunk 6 above)
Comment (gsganden) [reply creevey/load_funcs/image.py]: I'll add this note to the README.

File: creevey/load_funcs/image.py (hunk 5 above)
Comment (gsganden) [reply creevey/load_funcs/image.py]: Done

File: creevey/load_funcs/image.py (hunk 6 above)
Comment (gsganden) [reply creevey/load_funcs/image.py]: Done
